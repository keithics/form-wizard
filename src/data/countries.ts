export interface ICountry {
  name: string;
  value: string;
  currency: string;
  rate: number;
  isDefault?: boolean;
}

export const countries: ICountry[] = [
  { name: 'Hong Kong', value: 'hk', currency: 'HKD', rate: 1, isDefault: true },
  { name: 'United States of America', value: 'us', currency: 'USD', rate: 2 },
  { name: 'Australia', value: 'au', currency: 'AUD', rate: 3 },
];
