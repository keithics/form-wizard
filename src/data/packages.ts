import { IPackageValues } from 'components/forms/forms.interface';

export const packages: IPackageValues[] = [
  { value: 'standard', name: 'Standard', percentageAddOn: 0, isDefault: true },
  { value: 'sale', name: 'Sale', percentageAddOn: 50 },
  { value: 'safe', name: 'Super Safe', percentageAddOn: 75 },
];
