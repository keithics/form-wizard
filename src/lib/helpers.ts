export function getParams(): { [p: string]: string } {
  const urlSearchParams = new URLSearchParams(window.location.search);
  return Object.fromEntries(urlSearchParams.entries());
}
