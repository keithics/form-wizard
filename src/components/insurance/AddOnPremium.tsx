import { IPackageValues } from 'components/forms/forms.interface';
import { countries, ICountry } from 'data/countries';
import { calculateAddOn, DefaultKeys, getData } from 'lib/insurance.helpers';
import { IFormValues } from 'pages/Start';
import { Control, useWatch } from 'react-hook-form';

export function AddOnPremium({
  control,
  packageData,
  age,
}: {
  control: Control<IFormValues>;
  packageData: IPackageValues;
  age: number;
}) {
  const defaultCountry = getData(countries);

  const country = useWatch({
    control,
    name: 'country',
    defaultValue: defaultCountry.value,
  });

  // get country data object
  // const countryData = countries.find((c) => c.value === country) || countries[0];
  const countryData = getData(countries, DefaultKeys.VALUE, country) as ICountry;

  return age > 0 ? <>{calculateAddOn(countryData, packageData, age)}</> : null;
}
