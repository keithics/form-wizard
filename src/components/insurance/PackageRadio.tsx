import { IPackageValues, ISelect } from 'components/forms/forms.interface';
import { AddOnPremium } from 'components/insurance/AddOnPremium';
import { IFormValues } from 'pages/Start';
import { useFormContext, useWatch } from 'react-hook-form';

function PackageRadio({ name, values, required }: Omit<ISelect<IPackageValues>, 'register'>) {
  const { control, register } = useFormContext<IFormValues>();

  const age = useWatch({
    control,
    name: 'age',
    defaultValue: 0,
  });

  return (
    <>
      <div className="space-y-4 mt-10">
        {values.map((v) => (
          <div key={v.value} className="flex items-center">
            <input
              {...register(name, { required })}
              type="radio"
              id={v.name}
              defaultValue={v.value}
              className="input-radio"
              defaultChecked={v.isDefault}
            />
            <label htmlFor={v.name} className="labels-radio">
              {v.name} <AddOnPremium control={control} packageData={v} age={age} />
            </label>
          </div>
        ))}
      </div>
    </>
  );
}

export default PackageRadio;
