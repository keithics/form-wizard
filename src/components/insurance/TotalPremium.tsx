import { IPackageValues } from 'components/forms/forms.interface';
import { countries, ICountry } from 'data/countries';
import { packages } from 'data/packages';
import { calculatePremium, DefaultKeys, getData } from 'lib/insurance.helpers';
import { IFormValues } from 'pages/Start';
import { Control, useWatch } from 'react-hook-form';

export function TotalPremium({ control }: { control: Control<IFormValues> }) {
  const defaultCountry = getData(countries);
  const defaultPackage = getData(packages);

  const age = useWatch({
    control,
    name: 'age',
    defaultValue: 0,
  });

  const country = useWatch({
    control,
    name: 'country',
    defaultValue: defaultCountry.value,
  });

  const premium = useWatch({
    control,
    name: 'premium',
    defaultValue: defaultPackage.value,
  });

  // get country data object
  // const countryData = countries.find((c) => c.value === country) || countries[0];
  // const packageData = packages.find((p) => p.value === packageValue) || packages[0];

  const countryData = getData(countries, DefaultKeys.VALUE, country) as ICountry;
  const packageData = getData(packages, DefaultKeys.VALUE, premium) as IPackageValues;

  return age > 0 ? (
    <>
      <h3>
        Your premium is: {calculatePremium(countryData.rate, packageData.percentageAddOn, age)} {countryData.currency}
      </h3>
    </>
  ) : (
    <h3>Enter age to see your premium.</h3>
  );
}
