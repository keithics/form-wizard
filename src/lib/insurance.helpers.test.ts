import { expect } from '@jest/globals';
import { countries } from 'data/countries';
import { packages } from 'data/packages';
import { calculateAddOn, calculatePremium } from 'lib/insurance.helpers';

describe('Calculating Premiums', () => {
  describe('HK premium for 10years old ', () => {
    it('Should be 100 for regular', () => {
      expect(calculatePremium(1, 0, 10)).toMatchSnapshot();
    });

    it('Should be 150 for sale', () => {
      expect(calculatePremium(1, 50, 10)).toMatchSnapshot();
    });

    it('Should be 175 for super safe', () => {
      expect(calculatePremium(1, 75, 10)).toMatchSnapshot();
    });
  });

  describe('US premium for 10years old ', () => {
    it('Should be 200 for regular', () => {
      expect(calculatePremium(2, 0, 10)).toMatchSnapshot();
    });

    it('Should be 300 for sale', () => {
      expect(calculatePremium(2, 50, 10)).toMatchSnapshot();
    });

    it('Should be 350 for super safe', () => {
      expect(calculatePremium(2, 75, 10)).toMatchSnapshot();
    });
  });

  describe('AU premium for 10years old ', () => {
    it('Should be 300 for regular', () => {
      expect(calculatePremium(3, 0, 10)).toMatchSnapshot();
    });

    it('Should be 450 for sale', () => {
      expect(calculatePremium(3, 50, 10)).toMatchSnapshot();
    });

    it('Should be 525 for super safe', () => {
      expect(calculatePremium(3, 75, 10)).toMatchSnapshot();
    });
  });
});

describe('Calculating Premiums Addons', () => {
  describe('HK Addons for 10years old ', () => {
    it('Should be 0 for regular', () => {
      expect(calculateAddOn(countries[0], packages[0], 10)).toMatchSnapshot();
    });

    it('Should be 50 for sale', () => {
      expect(calculateAddOn(countries[0], packages[1], 10)).toMatchSnapshot();
    });

    it('Should be 75 for super safe', () => {
      expect(calculateAddOn(countries[0], packages[2], 10)).toMatchSnapshot();
    });
  });

  describe('US Addons for 10years old ', () => {
    it('Should be 0 for regular', () => {
      expect(calculateAddOn(countries[1], packages[0], 10)).toMatchSnapshot();
    });

    it('Should be 100 for sale', () => {
      expect(calculateAddOn(countries[1], packages[1], 10)).toMatchSnapshot();
    });

    it('Should be 150 for super safe', () => {
      expect(calculateAddOn(countries[1], packages[2], 10)).toMatchSnapshot();
    });
  });

  describe('AU Addons for 10years old ', () => {
    it('Should be 0 for regular', () => {
      expect(calculateAddOn(countries[2], packages[0], 10)).toMatchSnapshot();
    });

    it('Should be 150 for sale', () => {
      expect(calculateAddOn(countries[2], packages[1], 10)).toMatchSnapshot();
    });

    it('Should be 225 for super safe', () => {
      expect(calculateAddOn(countries[2], packages[2], 10)).toMatchSnapshot();
    });
  });
});
