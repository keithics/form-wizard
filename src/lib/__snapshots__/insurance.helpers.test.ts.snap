// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculating Premiums AU premium for 10years old  Should be 300 for regular 1`] = `"300.00"`;

exports[`Calculating Premiums AU premium for 10years old  Should be 450 for sale 1`] = `"450.00"`;

exports[`Calculating Premiums AU premium for 10years old  Should be 525 for super safe 1`] = `"525.00"`;

exports[`Calculating Premiums Addons AU Addons for 10years old  Should be 0 for regular 1`] = `""`;

exports[`Calculating Premiums Addons AU Addons for 10years old  Should be 150 for sale 1`] = `"+(150.00 AUD, 50%)"`;

exports[`Calculating Premiums Addons AU Addons for 10years old  Should be 225 for super safe 1`] = `"+(225.00 AUD, 75%)"`;

exports[`Calculating Premiums Addons HK Addons for 10years old  Should be 0 for regular 1`] = `""`;

exports[`Calculating Premiums Addons HK Addons for 10years old  Should be 50 for sale 1`] = `"+(50.00 HKD, 50%)"`;

exports[`Calculating Premiums Addons HK Addons for 10years old  Should be 75 for super safe 1`] = `"+(75.00 HKD, 75%)"`;

exports[`Calculating Premiums Addons US Addons for 10years old  Should be 0 for regular 1`] = `""`;

exports[`Calculating Premiums Addons US Addons for 10years old  Should be 100 for sale 1`] = `"+(100.00 USD, 50%)"`;

exports[`Calculating Premiums Addons US Addons for 10years old  Should be 150 for super safe 1`] = `"+(150.00 USD, 75%)"`;

exports[`Calculating Premiums HK premium for 10years old  Should be 100 for regular 1`] = `"100.00"`;

exports[`Calculating Premiums HK premium for 10years old  Should be 150 for sale 1`] = `"150.00"`;

exports[`Calculating Premiums HK premium for 10years old  Should be 175 for super safe 1`] = `"175.00"`;

exports[`Calculating Premiums US premium for 10years old  Should be 200 for regular 1`] = `"200.00"`;

exports[`Calculating Premiums US premium for 10years old  Should be 300 for sale 1`] = `"300.00"`;

exports[`Calculating Premiums US premium for 10years old  Should be 350 for super safe 1`] = `"350.00"`;
